drop package.json and fix CI

harrysarson opened the pull request:
This _might_ fix CI, we will see if it goes green :)

harrysarson committed:
drop package.json and fix CI

harrysarson committed:
fix travis by installing elm-format

We do not use elm-format but it seems like we still need
to install it!

dillonkearns commented:
Thanks for giving that a shot @harrysarson!

It looks like we'll need to use proper module names matching the files to get that CI check passing with 0.19.1. Which means we won't be able to use this same statement in this file here:

https://github.com/dillonkearns/elm-cli-options-parser/blob/master/examples/src/main.js

So we'll have to take a different approach for the example script runner.

harrysarson committed:
give examples some love

Install of lots of "Main" modules we now have lots of different
modules with their own names. Elm will compile them into a single
javascript file. The example js files then pick the module they
are interested in.

Should also include fixes for CI.

harrysarson commented:
I have tried a different approach. This is still running elm 0.19.0 mind but should make the switch easier. Will see if I can get travis' green tick.

harrysarson committed:
remove stray tab in travis config

harrysarson committed:
run elm make in subshell

We need to avoid perminantly changing the cwd.

harrysarson committed:
get success exit codes from examples

harrysarson committed:
fix example that require xmlhttprequest

harrysarson commented:
Green! That morphed into quite a large patch. I hope it is not too much to review, let me know if you want me to try and simplify the diff.

harrysarson commented:
Bump :)

dillonkearns commented:
Hello @harrysarson! So sorry for the delay! I've been bouncing between a few different things and wanted to give this a proper review. Thanks for the bump! And thanks again for working on this! ❤️

I really like this:

```elm
module.exports = (name, opts) => {
  const program = Elm[name].init({
```

Fantastic trick, and way cleaner!

A few things that I think would be nice to maintain from before these changes:

* Could we keep this index.html file that was removed in these spots here? https://github.com/dillonkearns/elm-cli-options-parser/pull/4/files#diff-5b0e3367cfabe5a0173e80753b827d6c https://github.com/dillonkearns/elm-cli-options-parser/pull/4/files#diff-465b86db411af86f4d4d9764eac664b1 - I've found it useful sometimes to maintain local copies in my repos of the code that I use for Ellie examples, so that I can make sure it's working locally and keep it in a compiling state as  I make changes. Then when I do a release, I can just copy-paste it to update the Ellie.
* I find it useful to have the core logic that's run in the travis in scripts (whether they're bash or NPM scripts). Could we keep the package.json as the source of truth for that logic like it was here? https://github.com/dillonkearns/elm-cli-options-parser/pull/4/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2 And then just run that from travis? Or is there a reason that wasn't working (sorry if I missed something there!)


Thanks again, these are really nice fixes! Let me know if there's anything that doesn't work with the suggestions I made there, and feel free to DM me on the Elm slack as well.

harrysarson commented:
> Could we keep this index.html file that was removed in these spots here?

Will do, I thought it was an auto generated one from the elm compiler checked in by accident. 

> I find it useful to have the core logic that's run in the travis in scripts (whether they're bash or NPM scripts). Could we keep the package.json as the source of truth for that logic like it was here? 

Looking back at this patch I agree. Will change.

dillonkearns commented:
Awesome, thank you so much!

harrysarson commented:
I moved the package.json file into the examples dir (because it is only needed when running examples). Should I move it back into the root of the repository?

dillonkearns commented:
I think it makes sense to have the `xhr` dependency only in the `examples` folder like you had it.

The way I like to structure things, I like to have the following in a root-level `package.json`:

* `npm test` runs the tests
* `elm`, `elm-test`, and `elm-format` are in devDependencies so that you can just `npm install` and have the right versions (and then your editors, npm scripts, etc. will be able to find the right version of each binary)
* Also nice to include the license and package name, but I guess not relevant here because we're not publishing an elm package

For the examples, it would be nice to have its own separate package.json. We could have the `approve` script in there and the `xhr` dependency. Then we could have an `npm run approve` script at the root-level package.json that just does `cd examples && npm install && npm run approve`. What do you think of that?

Then from travis, it only directly calls the scripts from the top level. That's what I usually do, but I'm open to other ideas, too!

harrysarson committed:
move core logic to package.json

harrysarson commented on package.json:
```diff
@@ -1,21 +1,9 @@
-    "build": "cd examples && elm make src/*.elm --output src/elm.js && cd -"
-  },
-  "author": "Dillon Kearns",
-  "license": "ISC",
```
Would you like me to add some of this metadata back into package.json?

dillonkearns replied:
The metadata isn't necessary (especially because the license type was incorrect). The changes to this file look good 👍

harrysarson commented on .travis.yml:
```diff
@@ -1,20 +1,18 @@
+  - ./examples/simple --name Steve
+  - ./examples/validation --name Steve
+  - ./examples/curl https://example.com
+  - echo "a\nb\nc" | ./examples/grep  b
```
These scripts (that actually run the examples and make sure they exit with a code of 0) are new in this PR. Should they be moved into package.json? The only thing is it becomes a very long line with lots of `&&`s if I do that.

dillonkearns replied:
Yeah, I think it will make it more maintainable to have them in a script, and then call that script from travis (rather than having it inline in the .travis.yml).

But a bash script would work, too! Then you don't need to squeeze everything onto one line. You can use `set -e` in the bash script to have the script exit if any of the lines in the script return non-0 exit status: https://stackoverflow.com/a/2871034.

harrysarson replied:
Agree!

harrysarson committed:
extract running of examples into approve.sh

harrysarson committed:
reinstate elm and elm-test as dev-deps

harrysarson commented:
Okay, maybe we are ready now!

dillonkearns commented on approve.sh:
```diff
@@ -0,0 +1,23 @@
+if [ ! -f elm.js ]; then
+    printf "Missing elm.js, run \`npm run build-examples\` to generate\n"
+    exit 1
+fi
```
```suggestion
npm run build-examples
```

Any reason not to do this?

harrysarson replied:
Slow? Although probably not too bad

dillonkearns replied:
I'm not sure I'm following what would make it slow, isn't it going to run it regardless? We can remove the `npm run build-examples` from the inline script in travis, and have it compile and run the examples in `approve.sh`.

Either way, though, robustness and foolproofing seems like an essential to me. I'm not sure the speed would be noticeable (unless I'm missing something).

harrysarson replied:
Agreed.

harrysarson committed:
always build elm.js beforing running examples

dillonkearns commented:
🎉 Yay! I can't thank you enough @harrysarson, thank you for taking on this big task and seeing it through 🙏 This is looking a lot cleaner than the original setup, I'm really like your improvements.
